backend/src/email/services/email-transactional.service.ts

```diff
@@ -74,17 +77,108 @@ async function sendMessage({
     recipients: [recipient],
     replyTo,
     attachments: sanitizedAttachments,
+    referenceId: emailMessageTransactionalId.toString(),
   }
   logger.info({
     message: 'Sending transactional email',
     action: 'sendMessage',
   })
-  const messageId = await EmailService.sendEmail(mailToSend)
+  const messageId = await EmailService.sendEmail(mailToSend, {
+    extraSmtpHeaders: { isTransactional: true },
+  })
   if (!messageId) {
     throw new Error('Failed to send transactional email')
   }
 }
 
+type CallbackMetaData = {
+  timestamp: Date
+  bounce?: {
+    bounceType: string
+    bounceSubType: string
+  }
+  complaint?: {
+    complaintFeedbackType: string
+    complaintSubType: string
+  }
+}
+async function handleStatusCallbacks(
+  type: string,
+  id: string,
+  metadata: CallbackMetaData
+): Promise<void> {
+  switch (type) {
+    case 'Delivery':
+      await EmailMessageTransactional.update(
+        {
+          status: TransactionalEmailMessageStatus.Delivered,
+          deliveredAt: metadata.timestamp,
+        },
+        {
+          where: { id },
+        }
+      )
+      break
+    case 'Bounce':
+      await EmailMessageTransactional.update(
+        {
+          status: TransactionalEmailMessageStatus.Bounced,
+          errorCode:
+            metadata.bounce?.bounceType === 'Permanent'
+              ? 'Hard bounce'
+              : 'Soft bounce',
+          errorSubType: metadata.bounce?.bounceSubType,
+        },
+        {
+          where: { id },
+        }
+      )
+      break
+    case 'Complaint':
+      await EmailMessageTransactional.update(
+        {
+          status: TransactionalEmailMessageStatus.Complaint,
+          errorCode: metadata.complaint?.complaintFeedbackType,
+          errorSubType: metadata.complaint?.complaintSubType,
+        },
+        {
+          where: { id },
+        }
+      )
+      break
+    case 'Open':
+      await EmailMessageTransactional.update(
+        {
+          status: TransactionalEmailMessageStatus.Opened,
+          openedAt: metadata.timestamp,
+        },
+        {
+          where: { id },
+        }
+      )
+      break
+    case 'Send':
+      await EmailMessageTransactional.update(
+        {
+          status: TransactionalEmailMessageStatus.Sent,
+          sentAt: metadata.timestamp,
+        },
+        {
+          where: { id },
+        }
+      )
+      break
+    default:
```

Comment:
Just thinking out loud (and aware that this might lead to over-engineering), but should we enforce some kind of order in the update operations?

For example, our code should disable updating a row whose status was previously set to `Opened` to `Sent`. (Can think of more examples.) This will prevent callbacks that come back in the wrong order (unlikely but possible?) from messing up our records?

Reply:
I thought about that:
- First thing is, there's no inherent order to these apart from just `Send` happens before the rest
- To implement this, it's actually not to determine an order of the statuses but rather just taking whichever has the latest timestamp. However, right now I see this problem to be quite minor, most of the time SES sends these events in the order of their timestamp (earliest to latest). We can come back to this if it becomes a problem but for now I don't think it's worth the effort

Reply:
The other ones I can think of is `Complaint` and `Bounce` should happen after `Delivery` (if at all), i.e. we wouldn't want `Delivery` to overwrite the previous two. It does seem like quite a lot of additional complexity to enforce this ordering though.

Will defer to you! (For stuff like this, do we add in the sprint tasks so that we can keep track of this and come back to in the future? Should these discussions be centralised somewhere so it's easier to track in the future? Feels like PR comments are not gr8 documentation)

Reply:
> The other ones I can think of is `Complaint` and `Bounce` should happen after `Delivery` (if at all), i.e. we wouldn't want `Delivery` to overwrite the previous two. It does seem like quite a lot of additional complexity to enforce this ordering though.
> 
> Will defer to you! (For stuff like this, do we add in the sprint tasks so that we can keep track of this and come back to in the future? Should these discussions be centralised somewhere so it's easier to track in the future? Feels like PR comments are not gr8 documentation)

Like I mentioned in the above comment, there's no inherent order to some of the events, if we're to implement we will need to do it based on the timestamps. I honestly don't think we need to tackle this, it's `medium effort, low-to-zero impact`